File: leetcode-js/221-maximum-square.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    'use strict';

    const dp = [];
    let edge = 0;

    for (let i = 0; i < matrix.length; i += 1) {
        dp[i] = [];
        for (let j = 0; j < matrix[0].length; j += 1) {
            dp[i][j] = parseInt(matrix[i][j]);
            if (i && j && dp[i][j]) {
                dp[i][j] = Math.min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1;
            }
            edge = Math.max(dp[i][j], edge);
        }
    }

    return edge * edge;
};

File: leetcode-java/221-maximum-square.java
public class Solution {
    public int maximalSquare(char[][] matrix) {
        if (matrix.length == 0) {
            return 0;
        }

        int m = matrix.length;
        int n = matrix[0].length;

        int[][] dp = new int[m + 1][n + 1];

        int edge = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    dp[i + 1][j + 1] = Math.min(dp[i][j], Math.min(dp[i + 1][j], dp[i][j + 1])) + 1;
                    edge = Math.max(edge, dp[i + 1][j + 1]);
                }
            }
        }

        return edge * edge;
    }
}
